### PR TITLE
close the underlying connection when the context is canceled

### LINKED ISSUE #2: respect the context
The Go standard library doesn't provide any way to use a `Context` when handshaking on pre-dialed TCP connection, see https://github.com/golang/go/issues/18482.

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #2.

### CODE REVIEW

**Review by Stebalien — CHANGES_REQUESTED** [review]
On return, we should *probably* replace the error with the context error if `ctx.Err()` is non-nil. Closing the connection will trigger some random error (write to closed connection, EOF, malformed something, etc.). However, that's not the *actual* error and callers usually expect `context.Canceled` or `context.DeadlineExceeded` here.

Otherwise, LGTM.

**Comment by marten-seemann** [comment]
@Stebalien Done. Please have another look.

File: transport.go [hunk 1]
```diff
@@ -43,23 +43,49 @@ var _ cs.Transport = &Transport{}
...
-	// TODO: use the ctx
-	// see https://github.com/golang/go/issues/18482
-	if err := cl.Handshake(); err != nil {
-		return nil, err
+	return t.handshake(ctx, insecure, cl)
+}
+
+func (t *Transport) handshake(
+	ctx context.Context,
+	// in Go 1.10, we need to close the underlying net.Conn
```
Comment (Stebalien) [inline transport.go:57]: I'm fine requiring go 1.11 but that's up to you (although we'd need to assert go-1.11 in that case with a build flag).

File: transport.go [hunk 2]
```diff
@@ -43,23 +43,49 @@ var _ cs.Transport = &Transport{}
...
+		case <-done:
+		case <-ctx.Done():
+			errChan <- ctx.Err()
+			insecure.Close()
+		}
+	}()
+
+	if err := tlsConn.Handshake(); err != nil {
+		// if the context was canceled, return the context error
+		errChan <- err
```
Comment (Stebalien) [inline transport.go]: This *works* but we can also just use:

```go
if ctxErr := ctx.Err(); ctxErr != nil {
    err = ctxErr
}
return nil, err
```

(no need for the channel dance)

**Review by Stebalien — APPROVED** [review]
Looks correct although the channel dance is probably unnecessary.

File: transport.go (hunk 1 above)
Comment (marten-seemann) [reply transport.go:57]: That would just exclude the file from builds with Go 1.10... Now that I already spent the time researching and fixing this, I'll leave it like this for now, and drop it once Go 1.12 is released.

File: transport.go (hunk 2 above)
Comment (marten-seemann) [reply transport.go]: Right, that's a lot nicer. Fixed.

File: transport.go [hunk 3]
```diff
@@ -43,23 +43,50 @@ var _ cs.Transport = &Transport{}
...
+		}
 		return nil, err
 	}
-	return t.setupConn(cl)
+	conn, err := t.setupConn(tlsConn)
+	if err != nil {
+		// if the context was canceled, return the context error
+		if ctxErr := ctx.Err(); ctxErr != nil {
+			return nil, ctxErr
+		}
```
Comment (Stebalien) [inline transport.go:87]: We still need to return the other error here, right?

**Review by Stebalien — CHANGES_REQUESTED** [review]

File: transport.go (hunk 3 above)
Comment (marten-seemann) [reply transport.go:87]: 😔yeah sure. Good catch.
Fixed.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**